src/Lock.sol

```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.24;
+
+import {Locker} from "./libraries/Locker.sol";
+
+/// @title Lock
+/// @notice A contract that provides a reentrancy lock for external calls
+/// @author Modified from https://github.com/Uniswap/universal-router/blob/main/contracts/base/Lock.sol
+contract Lock {
+    /// @notice Thrown when attempting to reenter a locked function from an external caller
+    error ContractLocked();
+
+    /// @notice Modifier enforcing a reentrancy lock that allows self-reentrancy
+    /// @dev If the contract is not locked, use msg.sender as the locker
+    modifier isNotLocked() {
+        // Apply a reentrancy lock for all external callers
```

Comment:
I find this a bit hard to follow all the cases.. could something like this be a little easier?

less else/if

```
address locker = _getLocker();
        if (locker != address(0)) {
            // Only allow self-reentrancy.
            if (msg.sender != address(this)) revert ContractLocked();
            _;
        } else {
            /// @dev For 7702 accounts the msg.sender is address(this) but we still need to set the lock if unset
            Locker.set(msg.sender);
            _;
            Locker.set(address(0));
        }
```

Reply:
Yea I like this